### PR TITLE
support channel balancer on datacoord

### PR DESCRIPTION
**Pull request by MrPresent-Han**
support channel balancer on datacoord to ensure channels distribution is even

#22324

### CODE REVIEW

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/datacoord/channel_manager.go [hunk 1]
```diff
@@ -246,37 +262,24 @@ func (c *ChannelManager) unwatchDroppedChannels() {
 	}
 }
 
-// NOT USED.
 func (c *ChannelManager) bgCheckChannelsWork(ctx context.Context) {
-	timer := time.NewTicker(bgCheckInterval)
+	timer := time.NewTicker(ChannelBalanceInterval)
```
Comment (congqixia) [inline internal/datacoord/channel_manager.go:266]: NIT: rename to `ticker` and add `defer ticker.Stop()`

File: internal/util/paramtable/component_param.go [hunk 2]
```diff
@@ -1356,6 +1360,14 @@ func (p *dataCoordConfig) initWatchTimeoutInterval() {
...
 }
 
+func (p *dataCoordConfig) initChannelBalanceSilentDuration() {
+	p.ChannelBalanceSilentDuration = time.Duration(p.Base.ParseInt64WithDefault("dataCoord.channel.balanceSilentDuration", 300)) * time.Second
+}
+
+func (p *dataCoordConfig) initChannelBalanceInterval() {
+	p.ChannelBalanceInterval = time.Duration(p.Base.ParseInt64WithDefault("dataCoord.channel.balanceInterval", 180)) * time.Second
+}
+
```
Comment (congqixia) [inline internal/util/paramtable/component_param.go:1370]: Config defined but not used

File: internal/datacoord/channel_manager.go (hunk 1 above)
Comment (soothing-rain) [inline internal/datacoord/channel_manager.go:266]: Not related in this PR but please add `defer timer.stop()` :)

File: internal/datacoord/channel_manager.go [hunk 3]
```diff
@@ -37,7 +37,10 @@ import (
 	"stathat.com/c/consistent"
 )
 
-const bgCheckInterval = 3 * time.Second
+const (
+	ChannelBalanceInterval       = 300 * time.Second
+	ChannelBalanceSilentDuration = 360 * time.Second
+)
```
Comment (congqixia) [inline internal/datacoord/channel_manager.go:43]: Use config instead

File: internal/datacoord/channel_manager.go [hunk 4]
```diff
@@ -863,3 +872,11 @@ func getReleaseOp(nodeID UniqueID, ch *channel) ChannelOpSet {
 	op.Add(nodeID, []*channel{ch})
 	return op
 }
+
+func (c *ChannelManager) IsSilent() bool {
```
Comment (soothing-rain) [inline internal/datacoord/channel_manager.go:876]: `IsSilent()` -> `isSilent`

File: internal/datacoord/channel_manager.go [hunk 5]
```diff
@@ -51,11 +54,14 @@ type ChannelManager struct {
 	assignPolicy     ChannelAssignPolicy
 	reassignPolicy   ChannelReassignPolicy
 	bgChecker        ChannelBGChecker
+	bgBalancePolicy  BalanceChannelPolicy
```
Comment (soothing-rain) [inline internal/datacoord/channel_manager.go:57]: bgBalancePolicy or balancePolicy?

File: internal/datacoord/channel_manager.go [hunk 6]
```diff
@@ -863,3 +872,11 @@ func getReleaseOp(nodeID UniqueID, ch *channel) ChannelOpSet {
 	op.Add(nodeID, []*channel{ch})
 	return op
 }
+
+func (c *ChannelManager) IsSilent() bool {
+	if c.stateTimer.hasRunningTimers() {
+		return false
+	}
+	lastSilentDuration := time.Since(c.lastActiveTimestamp)
+	return lastSilentDuration >= ChannelBalanceSilentDuration
```
Comment (congqixia) [inline internal/datacoord/channel_manager.go:881]: `ChannelBalanceSilentDuration` is a configurable duration. So it's just a guess there may be not incoming change?

File: internal/datacoord/policy.go [hunk 7]
```diff
@@ -363,6 +363,27 @@ func ConsistentHashDeregisterPolicy(hashRing *consistent.Consistent) DeregisterP
 	}
 }
 
+type BalanceChannelPolicy func(store ROChannelStore, ts time.Time) ChannelOpSet
+
+func AvgBalanceChannelPolicy(store ROChannelStore, ts time.Time) ChannelOpSet {
+	channelOps := make(ChannelOpSet, 0)
+	reAllocates, err := BgBalanceCheck(store.GetNodesChannels(), ts)
+	if err != nil {
+		log.Error("Failed to balance node channels", zap.Error(err))
```
Comment (soothing-rain) [inline internal/datacoord/policy.go:372]: nit: -> `failed`

File: internal/datacoord/channel_manager.go [hunk 8]
```diff
@@ -863,3 +872,11 @@ func getReleaseOp(nodeID UniqueID, ch *channel) ChannelOpSet {
 	op.Add(nodeID, []*channel{ch})
 	return op
 }
+
+func (c *ChannelManager) IsSilent() bool {
+	if c.stateTimer.hasRunningTimers() {
+		return false
+	}
+	lastSilentDuration := time.Since(c.lastActiveTimestamp)
```
Comment (soothing-rain) [inline internal/datacoord/channel_manager.go:880]: -> `return time.Since(c.lastActiveTimestamp) >= ChannelBalanceSilentDuration`

**Pull request merged**